File: ios/TasteByteERP/Features/HR/HRViewModel.swift
import Foundation

@MainActor
final class HRViewModel: ObservableObject {
    @Published var employees: [Employee] = []
    @Published var todayAttendance: Attendance?
    @Published var isClockedIn = false
    @Published var isLoading = false
    @Published var isClocking = false
    @Published var errorMessage: String?
    @Published var successMessage: String?
    @Published var searchText = ""
    @Published var currentTime = Date()

    private var timer: Timer?

    var filteredEmployees: [Employee] {
        guard !searchText.isEmpty else { return employees }
        let query = searchText.lowercased()
        return employees.filter {
            $0.fullName.lowercased().contains(query) ||
            $0.employeeNumber.lowercased().contains(query) ||
            ($0.department?.lowercased().contains(query) ?? false)
        }
    }

    func startClock() {
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.currentTime = Date()
            }
        }
    }

    func stopClock() {
        timer?.invalidate()
        timer = nil
    }

    func loadTodayAttendance() async {
        do {
            let response: APIResponse<Attendance> = try await APIClient.shared.get(
                APIEndpoints.attendanceToday
            )
            if response.success, let data = response.data {
                todayAttendance = data
                isClockedIn = data.clockIn != nil && data.clockOut == nil
            } else {
                todayAttendance = nil
                isClockedIn = false
            }
        } catch {
            todayAttendance = nil
            isClockedIn = false
        }
    }

    func clockIn() async {
        isClocking = true
        errorMessage = nil
        successMessage = nil

        do {
            let response: APIResponse<ClockInResponse> = try await APIClient.shared.postNoBody(
                APIEndpoints.clockIn
            )
            if response.success {
                isClockedIn = true
                successMessage = "Successfully clocked in!"
                await loadTodayAttendance()
            } else {
                errorMessage = response.error ?? "Failed to clock in"
            }
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to clock in"
        }

        isClocking = false
    }

    func clockOut() async {
        isClocking = true
        errorMessage = nil
        successMessage = nil

        do {
            let response: APIResponse<ClockOutResponse> = try await APIClient.shared.postNoBody(
                APIEndpoints.clockOut
            )
            if response.success {
                isClockedIn = false
                successMessage = "Successfully clocked out!"
                await loadTodayAttendance()
            } else {
                errorMessage = response.error ?? "Failed to clock out"
            }
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to clock out"
        }

        isClocking = false
    }

    func loadEmployees() async {
        isLoading = true
        errorMessage = nil

        do {
            let response: APIResponse<PaginatedResponse<Employee>> = try await APIClient.shared.getPaginated(
                APIEndpoints.employees,
                perPage: 100
            )
            if response.success, let paginated = response.data {
                employees = paginated.items
            } else {
                errorMessage = response.error ?? "Failed to load employees"
            }
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to load employees"
        }

        isLoading = false
    }
}

File: android/app/src/main/java/com/tastebyte/erp/features/hr/HrViewModel.kt
package com.tastebyte.erp.features.hr

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tastebyte.erp.core.network.ApiClient
import com.tastebyte.erp.models.Attendance
import com.tastebyte.erp.models.Employee
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class AttendanceState(
    val todayAttendance: Attendance? = null,
    val isClockedIn: Boolean = false,
    val isLoading: Boolean = false,
    val isClocking: Boolean = false,
    val error: String? = null,
    val successMessage: String? = null
)

data class EmployeeListState(
    val employees: List<Employee> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val searchQuery: String = ""
)

class HrViewModel : ViewModel() {

    private val _attendanceState = MutableStateFlow(AttendanceState())
    val attendanceState: StateFlow<AttendanceState> = _attendanceState.asStateFlow()

    private val _employeeListState = MutableStateFlow(EmployeeListState())
    val employeeListState: StateFlow<EmployeeListState> = _employeeListState.asStateFlow()

    init {
        loadTodayAttendance()
    }

    fun loadTodayAttendance() {
        viewModelScope.launch {
            _attendanceState.value = _attendanceState.value.copy(isLoading = true, error = null)
            try {
                val response = ApiClient.getService().getTodayAttendance()
                if (response.success && response.data != null) {
                    val attendance = response.data
                    _attendanceState.value = AttendanceState(
                        todayAttendance = attendance,
                        isClockedIn = attendance.clockIn != null && attendance.clockOut == null,
                        isLoading = false
                    )
                } else {
                    _attendanceState.value = AttendanceState(
                        isClockedIn = false,
                        isLoading = false
                    )
                }
            } catch (e: Exception) {
                _attendanceState.value = AttendanceState(
                    isLoading = false,
                    error = e.message ?: "Failed to load attendance"
                )
            }
        }
    }

    fun clockIn() {
        viewModelScope.launch {
            _attendanceState.value = _attendanceState.value.copy(
                isClocking = true,
                error = null,
                successMessage = null
            )
            try {
                val response = ApiClient.getService().clockIn()
                if (response.success && response.data != null) {
                    _attendanceState.value = _attendanceState.value.copy(
                        todayAttendance = response.data,
                        isClockedIn = true,
                        isClocking = false,
                        successMessage = "Clocked in successfully"
                    )
                } else {
                    _attendanceState.value = _attendanceState.value.copy(
                        isClocking = false,
                        error = response.error ?: "Clock in failed"
                    )
                }
            } catch (e: Exception) {
                _attendanceState.value = _attendanceState.value.copy(
                    isClocking = false,
                    error = e.message ?: "Network error"
                )
            }
        }
    }

    fun clockOut() {
        viewModelScope.launch {
            _attendanceState.value = _attendanceState.value.copy(
                isClocking = true,
                error = null,
                successMessage = null
            )
            try {
                val response = ApiClient.getService().clockOut()
                if (response.success && response.data != null) {
                    _attendanceState.value = _attendanceState.value.copy(
                        todayAttendance = response.data,
                        isClockedIn = false,
                        isClocking = false,
                        successMessage = "Clocked out successfully"
                    )
                } else {
                    _attendanceState.value = _attendanceState.value.copy(
                        isClocking = false,
                        error = response.error ?: "Clock out failed"
                    )
                }
            } catch (e: Exception) {
                _attendanceState.value = _attendanceState.value.copy(
                    isClocking = false,
                    error = e.message ?: "Network error"
                )
            }
        }
    }

    fun loadEmployees() {
        viewModelScope.launch {
            _employeeListState.value = _employeeListState.value.copy(isLoading = true, error = null)
            try {
                val response = ApiClient.getService().listEmployees(
                    search = _employeeListState.value.searchQuery.ifBlank { null }
                )
                if (response.success && response.data != null) {
                    _employeeListState.value = _employeeListState.value.copy(
                        employees = response.data.items,
                        isLoading = false
                    )
                } else {
                    _employeeListState.value = _employeeListState.value.copy(
                        isLoading = false,
                        error = response.error ?: "Failed to load employees"
                    )
                }
            } catch (e: Exception) {
                _employeeListState.value = _employeeListState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Network error"
                )
            }
        }
    }

    fun onEmployeeSearchChanged(query: String) {
        _employeeListState.value = _employeeListState.value.copy(searchQuery = query)
        loadEmployees()
    }

    fun clearSuccessMessage() {
        _attendanceState.value = _attendanceState.value.copy(successMessage = null)
    }
}
